pkg/build/api/v1/types.go

```diff
@@ -163,6 +163,9 @@ type BuildSource struct {
 	// repository.
 	ContextDir string `json:"contextDir,omitempty" description:"specifies sub-directory where the source code for the application exists, allows for sources to be built from a directory other than the root of a repository"`
 
+	// DockerfileName is the name of the Dockerfile that will be used to build the Docker image.
+	DockerfileName string `json:"dockerfileName,omitempty" description:"specifies the name of the Dockerfile to use for building the Docker image"`
```

Comment:
> But going halfway does not make sense to me. 

Totally agree, but we already do that with BuildSource and *Strategy structs.

> It is used to provide a dockerfile (in string form) for a docker strategy build to use, either by itself, or in conjunction with a git repo that is also provided. (and it will override a dockerfile that exists in that git repo)

Unfortunately since this is done in `fetchSource` [method](https://github.com/openshift/origin/blob/e66f133f88238b55441bab4772ed2456c72c7f01/pkg/build/builder/source.go#L60) used with docker and sti builder you may end up with unexpected behavior, where the Dockerfile will be overwritten in source strategy, even though you did not expect that to happen. Having specialized structs defends you against similar problems, where certain fields might have unexpected behavior for different strategies. I'm even ok with repeating those fields as long as the end result will be predictable.